### PR TITLE
`feat`: Requested timestamps that are too far in the future should be rejected and charged the minimum fee.

### PR DESCRIPTION
**Pull request by dfinity-ryancroote**
This PR introduces a future timestamp check for rate requests. If the timestamp is too far in the future, it is rejected and the minimum fee will be charged.

### CODE REVIEW

File: src/xrc/src/api.rs [hunk 1]
```diff
@@ -356,19 +357,31 @@ async fn handle_cryptocurrency_pair(
     }
 
     if !utils::is_caller_privileged(&caller) {
+        let current_timestamp = env.time_secs();
+        let timestamp_is_in_future = timestamp.value > current_timestamp;
```
Comment (THLO) [inline src/xrc/src/api.rs:361]: I think we can be a bit more lenient: If the timestamp maps to the current minute, it is still okay:

`timestamp_is_in_future = timestamp.value / 60 > current_timestamp / 60`

Comment (THLO) [inline src/xrc/src/api.rs:361]: Does it make sense to allow a privileged canister to request a timestamp in the future?
If it results in simpler code, then the answer is probably "yes"...

File: src/xrc/src/api.rs [hunk 2]
```diff
@@ -476,18 +489,30 @@ async fn handle_crypto_base_fiat_quote_pair(
...
             ChargeOption::OutboundRatesNeeded(num_rates_needed)
         };
 
         env.charge_cycles(charge_cycles_option)?;
+        if timestamp_is_in_future {
+            return Err(errors::timestamp_is_in_future_error(
+                timestamp.value,
+                current_timestamp,
+            ));
+        }
```
Comment (THLO) [inline src/xrc/src/api.rs:514]: There is quite a bit of duplicate code here...

Comment (dfinity-ryancroote) [reply src/xrc/src/api.rs:514]: I'm going to resolve this in a separate PR.

File: src/xrc/src/api.rs [hunk 3]
```diff
@@ -565,18 +590,25 @@ fn handle_fiat_pair(
     env: &impl Environment,
     request: &GetExchangeRateRequest,
 ) -> Result<QueriedExchangeRate, ExchangeRateError> {
-    let timestamp = utils::get_normalized_timestamp(env, request);
+    let requested_timestamp = utils::get_normalized_timestamp(env, request);
     let current_timestamp = env.time_secs();
-    let result = with_forex_rate_store(|store| {
-        store.get(
-            timestamp,
+    let result = if requested_timestamp <= current_timestamp {
```
Comment (THLO) [inline src/xrc/src/api.rs:595]: Same here.

File: src/xrc/src/api/test.rs [hunk 4]
```diff
@@ -1527,3 +1527,106 @@ fn cached_rate_with_few_collected_rates_is_ignored_for_privileged_canister() {
...
+    use crate::{candid::OtherError, errors::TIMESTAMP_IS_IN_FUTURE_ERROR_CODE, ONE_MINUTE};
+
+    use super::*;
+
+    /// This function tests that a crypto pair request with a timestamp in the future
+    /// is rejected and charged the minimum fee.
+    #[test]
+    fn handle_cryptocurrency_pair() {
+        let current_timestamp: u64 = 1678752000;
+        let future_timestamp = current_timestamp.saturating_add(2 * ONE_MINUTE);
```
Comment (THLO) [inline src/xrc/src/api/test.rs:1542]: Adding one minute would be enough, right?

File: src/xrc/src/api/test.rs [hunk 5]
```diff
@@ -1527,3 +1527,106 @@ fn cached_rate_with_few_collected_rates_is_ignored_for_privileged_canister() {
...
+            matches!(result, Err(ExchangeRateError::Other(OtherError { code, description: _ })) if code == TIMESTAMP_IS_IN_FUTURE_ERROR_CODE)
+        );
+    }
+
+    /// This function tests that a crypto/fiat pair request with a timestamp in the future
+    /// is rejected and charged the minimum fee.
+    #[test]
+    fn handle_crypto_base_fiat_quote_pair() {
+        let current_timestamp: u64 = 1678752000;
+        let future_timestamp = current_timestamp.saturating_add(2 * ONE_MINUTE);
```
Comment (THLO) [inline src/xrc/src/api/test.rs:1574]: Same here.

File: src/xrc/src/api/test.rs [hunk 6]
```diff
@@ -1527,3 +1527,106 @@ fn cached_rate_with_few_collected_rates_is_ignored_for_privileged_canister() {
...
+            matches!(result, Err(ExchangeRateError::Other(OtherError { code, description: _ })) if code == TIMESTAMP_IS_IN_FUTURE_ERROR_CODE)
+        );
+    }
+
+    /// This function tests that a fiat pair request with a timestamp in the future
+    /// is rejected and charged the minimum fee.
+    #[test]
+    fn handle_fiat_pair() {
+        let current_timestamp: u64 = 1678752000;
+        let future_timestamp = current_timestamp.saturating_add(2 * ONE_MINUTE);
```
Comment (THLO) [inline src/xrc/src/api/test.rs:1606]: And here.   :-)

**Review by THLO — APPROVED** [review]

File: src/xrc/src/api.rs (hunk 1 above)
Comment (dfinity-ryancroote) [reply src/xrc/src/api.rs:363]: The code is simpler, but it would not be much too much work. When I work on the duplicate code, I'll address this as well.

Comment (dfinity-ryancroote) [reply src/xrc/src/api.rs]: I don't think this is needed as `timestamp.value` is a normalized timestamp to the latest minute, so if it is still greater than the `current_timestamp` then `timestamp` is in the future.

File: src/xrc/src/api.rs (hunk 3 above)
Comment (dfinity-ryancroote) [reply src/xrc/src/api.rs]: `requested_timestamp` is normalized to the latest minute. It should not be larger than the `current_timestamp` at this point.

File: src/xrc/src/api.rs (hunk 1 above)
Comment (THLO) [reply src/xrc/src/api.rs]: You are right. If the timestamp is normalized, the seconds are removed. If it is still larger than current_timestamp, it must be the next minute or later. So, dividing by 60 is not necessary.

**Pull request merged**